src/lvalue.rs

```diff
@@ -206,9 +208,29 @@ impl<'a, 'tcx> EvalContext<'a, 'tcx> {
             _ => bug!("field access on non-product type: {:?}", base_layout),
         };
 
-        // FIXME(solson)
-        let base = self.force_allocation(base)?;
-        let (base_ptr, base_extra) = base.to_ptr_and_extra();
+        let (base_ptr, base_extra) = match base {
+            Lvalue::Ptr { ptr, extra } => (ptr, extra),
+            Lvalue::Local { frame, local, field } => match self.stack[frame].get_local(local, field.map(|(i, _)| i)) {
+                Value::ByRef(ptr) => {
+                    assert!(field.is_none(), "local can't be ByRef and have a field offset");
+                    (ptr, LvalueExtra::None)
+                },
+                Value::ByVal(_) => {
+                    assert_eq!(offset.bytes(), 0, "ByVal can only have 1 non zst field with offset 0");
+                    return Ok(base);
+                },
+                Value::ByValPair(_, _) => {
+                    assert!(field_index < 2);
```

Comment:
We most likely can hit it with zst fields, but right now we don't create any ByValPairs for custom structs.